glean-core/python/glean/net/ping_upload_worker.py

```diff
@@ -75,65 +71,78 @@ def _test_process_sync(cls) -> bool:
-                serialized_ping = next(lines)
-                valid_content = True
-            except StopIteration:
-                valid_content = False
-        # On Windows, we must close the file before deleting it
-        if not valid_content:
-            path.unlink()
-            log.error("Invalid ping content in {}".format(path.resolve()))
-            return False
-    except FileNotFoundError:
-        log.error("Could not find ping file {}".format(path.resolve()))
-        return False
-    except IOError as e:
-        log.error("IOError when reading {}: {}".format(path.resolve(), e))
-        return False
-
-    processed = configuration.ping_uploader.do_upload(
-        url_path, serialized_ping, configuration
-    )
-
-    if processed:
-        path.unlink()
-        log.debug("{} was deleted".format(path.name))
-
-    return processed
+        parsed_headers = json.loads(headers_as_json)
+        for header_name, header_value in parsed_headers.items():
+            headers.append((header_name, header_value))
+    except json.decoder.JSONDecodeError as e:
+        log.error("Error while parsing headers for ping " + document_id)
+    finally:
+        return headers
+
+
+def _process(data_dir: Path, configuration) -> bool:
+
+    # Import here to avoid cyclical import
+    from ..glean import Glean
+
+    if not Glean.is_initialized():
+        # If we're in the uploader process and the Glean SDK was not initialized,
+        # initialize it in "upload-only" mode.
+        data_dir = ffi_support.new("char[]", _ffi.ffi_encode_string(str(data_dir)))
+        _ffi.lib.glean_initialize_upload_only(data_dir)
+
+    while True:
+        incoming_task = ffi_support.new("FfiPingUploadTask *")
+        _ffi.lib.glean_get_upload_task(incoming_task)
+
+        tag = incoming_task.tag
+        if tag == UploadTaskTag.UPLOAD:
+            # Ping data is available for upload: parse the structure but make
+            # sure to let Rust free the memory.
+            doc_id = _ffi.ffi_decode_string(
+                incoming_task.upload.document_id, free_memory=False
+            )
+            url_path = _ffi.ffi_decode_string(
+                incoming_task.upload.path, free_memory=False
+            )
+            body = _ffi.ffi_decode_byte_buffer(
+                incoming_task.upload.body, free_memory=False
+            )
+            headers = _ffi.ffi_decode_string(
+                incoming_task.upload.headers, free_memory=False
+            )
+
+            # Delegate the upload to the uploader.
+            upload_result = configuration.ping_uploader.do_upload(
+                url_path, body, _parse_ping_headers(headers, doc_id), configuration
+            )
+
+            # Process the response.
+            _ffi.lib.glean_process_ping_upload_response(
+                incoming_task, upload_result.to_ffi()
+            )
+        elif tag == UploadTaskTag.WAIT:
+            # TODO: what to do here?
```

Comment:
In Kotlin we delegate the waiting part to the `WorkManager`, here we don't have that. Since we are in a separate process, waiting that much might be an option.

Reply:
This task was supposed to eventually gain a timeout, so the uploader knows how long to wait.

Right now the only way `Wait` can be returned is if the directory is still scanned.
Given that the python uploader will trigger a re-scan of the directory every time (because it starts up a manager every time it's launched) it should probably wait a couple of times (say 2-3 times if wait is returned) and otherwise just end itself.

Reply:
Ok, so, if I got that right, @badboy , the `time.wait` is a reasonable strategy here.

Reply:
Yes -- and the `time.sleep` won't block the main process or threads, so is not a problem here.  Only detail I think missing here from @badboy's comment is that we should only wait *n* times before `return False` just to prevent the case of an upload process running forever if there is some sort of bug.

Reply:
Oh, good idea.